### ISSUE TITLE
Redirect functionality on invalid or no token in request

### ISSUE THREAD

**Issue by LPmaverick** [OWASP/www-project-csrfguard#215]
I understand that when a request comes into my application, against an endpoint I have protected, with either no token or an invalid token that this library will perform a redirect to my configured error page/endpoint.  

I was recently testing an application that was using v3.1.0 of this library and they had setup `/home` to be their configured error endpoint.  The functionality within this controller was to redirect to another (protected) endpoint within the application.  This internal redirect would work.  Upgrading the CSRFGuard library to the latest version causes this behavior to (as expected) result in an infinite loop.

Why did this functionality change between v3.1.0 and the current version?  Was it just a functional change or to remediate against a vulnerability?

**Comment by forgedhallpass** [comment]
Hello @LPmaverick,

I have not encountered this issue before. I believe I've usually forced a log-out when the CSRF token validation failed. Could you please provide the configurations on how to reproduce it on the project's test application?

p.s. in case you've found the issue and know how to fix it, I'd like to encourage you to provide a PR.

**Comment by LPmaverick** [comment]
Well, the library has already fixed this in the current version, but I have not checked other versions to nail down exactly which version fixed this.  I expected this flow to result in an infinite redirect loop; which is happening in the current version but not in v3.1.0.  

Here is the properties file. [csrfguard.zip](https://github.com/OWASP/www-project-csrfguard/files/12440268/csrfguard.zip)

Here is the xml configuration
```
<listener>
	   <listener-class>org.owasp.csrfguard.CsrfGuardHttpSessionListener</listener-class>
	</listener>
	<listener>
	   <listener-class>org.owasp.csrfguard.CsrfGuardServletContextListener</listener-class>
	</listener>
	<context-param>
	   <param-name>Owasp.CsrfGuard.Config</param-name>
	   <!-- Location of file within the app's file system -->
	   <param-value>WEB-INF/csrfguard-4.x.properties</param-value>
	</context-param>
	<filter>
	   <filter-name>CSRFGuard</filter-name>
	   <filter-class>org.owasp.csrfguard.CsrfGuardFilter</filter-class>
	</filter>
	<filter-mapping>
	   <filter-name>CSRFGuard</filter-name>
	   <url-pattern>/*</url-pattern>
	</filter-mapping>
	<servlet>
   <servlet-name>JavaScriptServlet</servlet-name>
	   <servlet-class>org.owasp.csrfguard.servlet.JavaScriptServlet</servlet-class>
	</servlet>
	<servlet-mapping>
	   <servlet-name>JavaScriptServlet</servlet-name>
	   <url-pattern>/JavaScriptServlet</url-pattern>
	</servlet-mapping>
```

Here is the home servlet
```
@WebServlet("/home")
public class HomeServlet extends HttpServlet {
	private static final long serialVersionUID = 1L;

	/**
	 * @see HttpServlet#doGet(HttpServletRequest request, HttpServletResponse
	 *      response)
	 */
	protected void doGet(HttpServletRequest request, HttpServletResponse response)
			throws ServletException, IOException {
		System.out.println("Reached home servlet");
		response.sendRedirect("CentralController");
	}
}
```

Here is the protected redirect servlet
```
@WebServlet("/CentralController")
public class CentralController extends HttpServlet {
	final String PAGE_NAME = "hPageName";
	final String ACTION_NAME = "hActionName";

	/**
	 * @see HttpServlet#doGet(HttpServletRequest request, HttpServletResponse
	 *      response)
	 */
	protected void doGet(HttpServletRequest req, HttpServletResponse response) throws ServletException, IOException {
		req.getServletContext().getRequestDispatcher("/WEB-INF/jsp/home.jsp").forward(req, response);
	}
}
```

To trigger, you can access either `/home` or `/CentralController`

**Comment by forgedhallpass** [comment]
> Well, the library has already **fixed this in the current version**, but I have not checked other versions to nail down exactly which version fixed this. I expected this flow to result in an infinite redirect loop; which is **happening in the current version** but not in v3.1.0.

@LPmaverick I am not sure I understand. Your "fixed this in the current version" and "infinite redirect loop happening in the current version" statements seem to be contradicting each other.

Could you please clarify if the issue is present in the LATEST (4.3.0) version of the solution?

**Comment by LPmaverick** [comment]
So, when I first came across the application's configuration, what I expected to happen was an infinite redirect loop because the unprotected home endpoint was redirecting to the protected controller endpoint.  This was all happening server-side (java) which meant that the auto-token injection provided by the JavaScriptServlet wouldn't happen.  

I tested their configuration against a test app we have for such cases, which uses csrfguard v4.1.2 (my mistake I thought I was using the most current version), and I got the expected result.  However, when testing the application itself (using csrfguard v3.1.0), I did not experience an infinite loop.  Instead, whether I hit the home or controller endpoint, I would end up being served the jsp page delivered by the controller servlet with the token appended in the browser URL.  So, I changed my test app to use v3.1.0 and I was able to replicate their result which told me that it wasn't something the application was doing but rather the csrfguard library itself.

What I am ultimately trying to determine is, which version behavior is correct?
Is the behavior of v3.1.0 is a potential vulnerability?
If v4.x is the correct behavior, then the "fix" is to inform the application support team to upgrade their csrfguard version to at least v4.1.2 (known good).

**Comment by LPmaverick** [comment]
Hope I am making sense.

**Comment by LPmaverick** [comment]
Just had a javax servlet 2.5 application using csrfguard v4.1.2 not trigger the infinite redirect loop as I would have expected it to.

**Comment by forgedhallpass** [comment]
Hello @LPmaverick,

1. I am still not sure what you would like to achieve here. In the code above you are setting up the forward and the redirect unrelated to CSRFGuard, and in the example properties file you have two unprotected pages that both reference the same path:

```properties
org.owasp.csrfguard.unprotected.Home=%servletContext%/home
org.owasp.csrfguard.unprotected.GoodEndpoint=%servletContext%/home
```

2. Besides, when using the latest version of the application (especially if there was a major version change), please use the [latest properties file](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard/src/main/resources/csrfguard.properties) as well, just to be on the safe side. This also contains descriptions of the configurable attributes and their behavior.

3. The Redirect action can be found [here](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard/src/main/java/org/owasp/csrfguard/action/Redirect.java#L48), and it's only invoked if there was potential CSRF attack detected.

4. CSRFGuard is open-source, hence you should be able to easily debug exactly what is happening. 

5. Previous versions of CSRFGuard are not supported anymore and it is recommended to upgrade to the latest version.

**Comment by LPmaverick** [comment]
I agree with you on upgrading to the latest version of the library.  However, what I have brought to you is not code I support but rather other's code that I am consulting on.  What they are experiencing is not what I expected they should experience, nor is it what I have tested and experienced.

What I would like to achieve is to understand why CSRFGuard is behaving differently based on the code snippets I have provided.  If an application must have GET endpoints protected, they must also setup their redirect page/endpoint to be unprotected, correct?  Additionally, if their redirect page/endpoint's only job is to redirect to their application's "main/home" endpoint, then that endpoint as well must also be setup as unprotected, correct?  What configuration property would make CSRFGuard append the token to a redirect being triggered by the application's back-end servlet?
